### PR TITLE
Made Nearby show all pins that could be presented on the screen, rather than a circle

### LINKED ISSUE #5480: Make Nearby show all pins that could be presented on the screen, rather than a circle
Currently Nearby shows pins in a way that often does not cover the whole screen:

At least at default zoom level, I feel that it should be possible to load more pins, enough to fill all areas of the screen.

Similar to how [Wikishootme](https://wikishootme.toolforge.org/#lat=48.869598169524025&lng=2.3070538043975835&zoom=16&layers=wikidata_image,wikidata_no_image) loads pins:
Screenshot 2024-01-27 at 12 23 33

Wikishootme loads rather fast, our query is more complex but it is worth trying. We can also speed it up with some trick like #4560

### PR DESCRIPTION
**Pull request by kanahia1**
**Description (required)**

Fixes #5480

What changes did you make and why?

**Tests performed (required)**

Tested prodDebug on Samsung S21 FE with API level 33.

**Screenshots (for UI changes only)**

https://github.com/commons-app/apps-android-commons/assets/114223204/f78aa3fd-f346-4da5-9261-700bfe603abd

### CODE REVIEW

File: app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java [hunk 1]
```diff
@@ -328,6 +329,76 @@ public List<Place> getNearbyPlaces(final LatLng cur, final String language, fina
...
+        final String query = wikidataQuery
+            .replace("${LAT_WEST}", String.format(Locale.ROOT, "%.4f", westCornerLat))
+            .replace("${LONG_WEST}", String.format(Locale.ROOT, "%.4f", westCornerLong))
+            .replace("${LAT_EAST}", String.format(Locale.ROOT, "%.4f", eastCornerLat))
+            .replace("${LONG_EAST}", String.format(Locale.ROOT, "%.4f", eastCornerLong))
+            .replace("${LANG}", language);
+        final HttpUrl.Builder urlBuilder = HttpUrl
+            .parse(sparqlQueryUrl)
+            .newBuilder()
+            .addQueryParameter("query", query)
```
Comment (shashankiitbhu) [inline app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java:375]: can you replace this with a safer (null safety) version like below:

```
final HttpUrl.Builder urlBuilder = Objects.requireNonNull(HttpUrl
                .parse(sparqlQueryUrl))
            .newBuilder()
            .addQueryParameter("query", query)
            .addQueryParameter("format", "json");
 ```

**Comment by kanahia1** [comment]
@nicolas-raoul, Can you please review this PR 🙂

File: app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java [hunk 2]
```diff
@@ -328,6 +329,76 @@ public List<Place> getNearbyPlaces(final LatLng cur, final String language, fina
...
+     * @param language             The language for the query.
+     * @param shouldQueryForMonuments  Flag indicating whether to include monuments in the query.
+     * @param customQuery          Optional custom SPARQL query to use instead of default queries.
+     * @return                     A list of nearby places.
+     * @throws Exception           If an error occurs during the retrieval process.
+     */
+    @Nullable
+    public List<Place> getNearbyPlaces(
+        final fr.free.nrw.commons.location.LatLng screenTopRight,
+        final fr.free.nrw.commons.location.LatLng screenBottomLeft,final String language,
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java:346]: please insert a space after each `,` comma, here and in the rest of the PR (many occurrences)

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyController.java [hunk 3]
```diff
@@ -130,6 +129,90 @@ public NearbyPlacesInfo loadAttractionsFromLocation(final LatLng curLatLng, fina
...
+
+        if (searchLatLng == null) {
+            Timber.d("Loading attractions nearby, but curLatLng is null");
+            return null;
+        }
+
+        List<Place> places = nearbyPlaces.getFromWikidataQuery(screenTopRight,screenBottomLeft,Locale.getDefault().getLanguage(),shouldQueryForMonuments, customQuery);
+
+        if (null != places && places.size() > 0) {
+            LatLng[] boundaryCoordinates = {places.get(0).location,   // south
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyController.java:163]: Please put `places.get(0).location,   // south` on its own line

File: app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java [hunk 4]
```diff
@@ -1153,11 +1154,36 @@ public void listOptionMenuItemClicked() {
 
     @Override
     public void populatePlaces(final fr.free.nrw.commons.location.LatLng curlatLng) {
-        if (curlatLng.equals(getLastMapFocus())) { // Means we are checking around current location
-            populatePlacesForCurrentLocation(getLastMapFocus(), curlatLng, null);
-        } else {
-            populatePlacesForAnotherLocation(getLastMapFocus(), curlatLng, null);
+        IGeoPoint screenTopRight = mapView.getProjection().fromPixels(mapView.getWidth(), 0);
+        IGeoPoint screenBottomLeft = mapView.getProjection().fromPixels(0, mapView.getHeight());
+        fr.free.nrw.commons.location.LatLng screenTopRightLatLng = new fr.free.nrw.commons.location.LatLng(screenBottomLeft.getLatitude(),screenBottomLeft.getLongitude(),0);
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java:1159]: Please break down very long lines.

File: app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java [hunk 5]
```diff
@@ -1153,11 +1154,36 @@ public void listOptionMenuItemClicked() {
...
+        fr.free.nrw.commons.location.LatLng screenBottomLeftLatLng = new fr.free.nrw.commons.location.LatLng(screenTopRight.getLatitude(),screenTopRight.getLongitude(),0);
+        fr.free.nrw.commons.location.LatLng emptyLatLng = new fr.free.nrw.commons.location.LatLng(0.0,0.0,0);
+
+        if (screenTopRightLatLng.getLatitude() == 0.0 && screenTopRightLatLng.getLongitude() == 0.0 && screenBottomLeftLatLng.getLatitude() == 0.0 && screenBottomLeftLatLng.getLongitude() == 0.0){
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java:1163]: Can you add a comment explaining in which case that would typically happen?
Also, comparing doubles can be tricky, see https://stackoverflow.com/a/25160509/226958

File: app/src/main/resources/queries/nearby_query.rq [hunk 6]
```diff
@@ -13,10 +13,10 @@ SELECT
   (SAMPLE(?commonsCategory) AS ?commonsCategory)
 WHERE {
   # Around given location
-  SERVICE wikibase:around {
+  SERVICE wikibase:box {
     ?item wdt:P625 ?location.
-    bd:serviceParam wikibase:center "Point(${LONG} ${LAT})"^^geo:wktLiteral. # Longitude latitude
-    bd:serviceParam wikibase:radius "${RAD}". # Radius in kilometers.
+     bd:serviceParam wikibase:cornerWest "Point(${LONG_WEST} ${LAT_WEST})"^^geo:wktLiteral.
+     bd:serviceParam wikibase:cornerEast "Point(${LONG_EAST} ${LAT_EAST})"^^geo:wktLiteral.
```
Comment (nicolas-raoul) [inline app/src/main/resources/queries/nearby_query.rq:19]: Great! :-)

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
It works and I love it already!

Just a few nits to correct.

File: app/src/main/resources/queries/nearby_query.rq (hunk 6 above)
Comment (kanahia1) [reply app/src/main/resources/queries/nearby_query.rq:19]: Thanks :-)

File: app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java (hunk 5 above)
Comment (kanahia1) [reply app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java]: Sure, I have added documentation in newer commit https://github.com/commons-app/apps-android-commons/pull/5553/commits/4d9b10b90536d56e83e820ede39b01ec19aa5577

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyController.java [hunk 7]
```diff
@@ -58,13 +58,12 @@ public NearbyController(NearbyPlaces nearbyPlaces) {
      * @return NearbyPlacesInfo a variable holds Place list without distance information
      * and boundary coordinates of current Place List
      */
-    public NearbyPlacesInfo loadAttractionsFromLocation(final LatLng curLatLng, final LatLng searchLatLng,
+    public NearbyPlacesInfo loadAttractionsFromLocation(final LatLng curLatLng,final LatLng searchLatLng,
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyController.java:61]: Space after comma

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyController.java [hunk 8]
```diff
@@ -130,6 +130,91 @@ public NearbyPlacesInfo loadAttractionsFromLocation(final LatLng curLatLng, fina
...
+
+        Timber.d("Loading attractions near %s", searchLatLng);
+        NearbyPlacesInfo nearbyPlacesInfo = new NearbyPlacesInfo();
+
+        if (searchLatLng == null) {
+            Timber.d("Loading attractions nearby, but curLatLng is null");
+            return null;
+        }
+
+        List<Place> places = nearbyPlaces.getFromWikidataQuery(screenTopRight,screenBottomLeft,Locale.getDefault().getLanguage(),shouldQueryForMonuments, customQuery);
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyController.java:161]: spaces, line length

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java [hunk 9]
```diff
@@ -95,4 +95,23 @@ public List<Place> getFromWikidataQuery(final LatLng cur, final String lang,
...
+     * @param screenBottomLeft     The bottom left corner of the screen (latitude, longitude).
+     * @param lang                 The language for the query.
+     * @param shouldQueryForMonuments  Flag indicating whether to include monuments in the query.
+     * @param customQuery          Optional custom SPARQL query to use instead of default queries.
+     * @return                     A list of places obtained from the Wikidata query.
+     * @throws Exception           If an error occurs during the retrieval process.
+     */
+    public List<Place> getFromWikidataQuery(
+        final fr.free.nrw.commons.location.LatLng screenTopRight,
+        final fr.free.nrw.commons.location.LatLng screenBottomLeft,final String lang,final boolean shouldQueryForMonuments,
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java:112]: spaces

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java [hunk 10]
```diff
@@ -95,4 +95,23 @@ public List<Place> getFromWikidataQuery(final LatLng cur, final String lang,
...
+        @Nullable final String customQuery) throws Exception {
+        return okHttpJsonApiClient
+            .getNearbyPlaces(screenTopRight,screenBottomLeft,lang,shouldQueryForMonuments, customQuery);
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java:115]: spaces

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
A few more very minor things

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java (hunk 10 above)
Comment (kanahia1) [reply app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java]: Sure !, done in the newer commit

File: app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java [hunk 11]
```diff
@@ -328,6 +329,76 @@ public List<Place> getNearbyPlaces(final LatLng cur, final String language, fina
         throw new Exception(response.message());
     }
 
+    /**
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java:332]: Do we still need the methods above and below this one (with the `radius` parameter)?

Comment (kanahia1) [reply app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java:332]: I think it's required during upload to prepopulate categories and depictions based on the image coordinates where their `radius` is used.

Comment (kanahia1) [reply app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java:330]: Sure!

**Comment by kanahia1** [comment]
@nicolas-raoul, I have made changes suggested by you https://github.com/commons-app/apps-android-commons/pull/5553/commits/1e732edb78fe5bb11059d0df58c6c4653759d2c3. I completely forgot about the monuments since they are also using radius query we may need different files for them, Can you please share the names for `nearby_query_monuments.rq`

**Comment by nicolas-raoul** [comment]
How about rectangle_query_for_nearby_monuments.rq?
Is a radius query needed for monuments?

**Comment by kanahia1** [comment]
I found out that everywhere for radius, shouldQueryForMonuments is set to false. So removing radius query for monuments will be okay :-)

**Comment by kanahia1** [comment]
Hey @nicolas-raoul 

```
Caused by: com.android.build.gradle.tasks.ResourceException: /home/runner/work/apps-android-commons/apps-android-commons/app/src/main/res/values-yue-hant: Error: Invalid resource directory name
	at com.android.build.gradle.tasks.MergeResources.doFullTaskAction(MergeResources.kt:291)
	at com.android.build.gradle.tasks.MergeResources.doTaskAction(MergeResources.kt:320)
```

Is failing of checks related to https://github.com/commons-app/apps-android-commons/issues/5571

**Comment by nicolas-raoul** [comment]
Sorry for the delay, would you mind fixing thee conflict?

**Comment by kanahia1** [comment]
Hey @nicolas-raoul, I think all the conflicts are fixed now. Can you please look into it.

**Comment by kanahia1** [comment]
Friendly ping @nicolas-raoul

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Would you mind performing the stylistic changes in a different pull request?
They make it very difficult to spot what actually changed.
Thanks! :-)

**Comment by kanahia1** [comment]
@nicolas-raoul, Sure, I will do it!

**Comment by kanahia1** [comment]
Hey @nicolas-raoul, Sorry to ask you a simple question but can you please give more information what kind of `stylistic` changes should I make or should do another pull request by closing this one (So that I could be easy to review)

**Comment by nicolas-raoul** [comment]
I must have dreamt, I don't see them anymore, sorry my bad! 😰

**Review by nicolas-raoul — APPROVED** [review]
Working well even for Mannhattan, thanks!

**Pull request merged**